### ISSUE TITLE
Physics - Collision Filtering

### ISSUE THREAD

**Issue by Jase0000** [Scirra/Construct-feature-requests#68]
### Reviewed guidelines

- [X] I have read and understand the suggestion guidelines

### Checked for duplicate suggestions

- [X] I checked for existing similar suggestions

### Summary

Currently, many ideas are impossible to achieve due to lack of collision filtering with the Physics behaviour.

This includes:

1. Having a platformer game with 2 layers of objects where they technically exist in similar positions but on different layers, cannot allow the physics to continue behaving otherwise they will interact with each other between layers.
2. Having physics vehicles where you want to drive around a loop-de-loop, a wheel may need to collide at specific times whilst the other wheel must not collide. If you workaround this with 2 separate wheel objects, then the next issue is multiple vehicles for a multiplayer game.
3. Say you have a table and a lettuce, you cannot have some lettuces on the table whilst some are on the floor rolling past the base of the table.
4. Some possible 3D implementations of using physics, if objects are on different Z elevations, they are technically overlapping and therefore collide, even if you do the complex stuff of detecting whether they are on similar Z elevations and such.

### Possible workarounds or alternatives

I have explored this valiently, and did not succeed.
I tried adding 2 physics behaviours to 1 object and setting each behaviour to collide/not collide with specific objects (thinking this will separate the objects, but it does not).

The closest possible workaround is a nightmare to consider, which is a hidden parallel universe of all objects far offscreen, all cloned objects so not same instances, and then positioning the real-world objects according to the results from these parallel universes.

### Proposed solution

I empathise that the current implementation is sort of "stuck" at the moment beyond Scirra's control.

I wonder if there's a new implementation of Box2D/ASM.JS that could be used to update the physics.

If not, perhaps an entirely new Physics plugin could be developed, possibly labelled differently such as "Advance Physics" or forego the word "Physics" to not confuse beginners (and if beginners are trying to do complex collision filtering, they will quickly find this new alternative plugin).

I also have the feeling this could be beneficial in other ways, such as granting new physics welds/joints and such (if the current implementation is out-of-date and does not have these).

### Why is this idea important?

On a personal level, there is a huge game idea I wish to do, but cannot begin until there is some form of collision filtering for physics - I had explored workarounds but cannot solve this.

Physics is extremely common for games, some games running entirely on physics including the player and their movement. More power and control over this would be GREATLY appreciated.

**Comment by TackerTacker** [comment]
I boiled it down to one change that could be made to enable collision filtering in C3.
Enable/disable collisions should be applied to instances instead of globally to the object. 
![grafik](https://github.com/Scirra/Construct-feature-requests/assets/18266300/83d0f957-bb55-41a0-b532-282de2c01bb3)
I don't know how difficult it is to implement this, but I think it is at least more realistic than an entire new physics engine.

**Comment by Nabu0001** [comment]
This is very important. Every physics engine has "per instance" collision filtering. In Box2D, there are properties called 'categoryBits',  'maskBits' and 'groupIndex' with this purpose. And it was a main feature of the Chipmunk plugin under C2.

Please, consider adding this absolutely needed feature. Best regards to Construct Team and everyone.

**Comment by Jase0000** [comment]
> I boiled it down to one change that could be made to enable collision filtering in C3. Enable/disable collisions should be applied to instances instead of globally to the object. ![grafik](https://private-user-images.githubusercontent.com/18266300/306692168-83d0f957-bb55-41a0-b532-282de2c01bb3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkwNzUzNzksIm5iZiI6MTcwOTA3NTA3OSwicGF0aCI6Ii8xODI2NjMwMC8zMDY2OTIxNjgtODNkMGY5NTctYmI1NS00MWEwLWI1MzItMjgyZGUyYzAxYmIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI3VDIzMDQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiYjc1Y2MxMzcxYmNjZjZhOWVhZDMxMzRmNDMxNzZhYmYyZjVkZmZhYjI3YjkyMDJkMjM0OTYwMGMwODA1NDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.0qhP4TQaQG7WKL4dO0CbcknNuoannzReOhezqXQSMxE) I don't know how difficult it is to implement this, but I think it is at least more realistic than an entire new physics engine.

Yes, this would be absolutely perfect - I seem to recall something about Scirra being stuck with whatever Box2D implementation they use and it has limited features, but for the life of me I cannot find a forum post mentioning this, I swear I saw it...

But yes, if it can be done in current Box2D implementation, then yes yes yesss!

**Comment by Wackytoaster** [comment]
Yes please. Here's a practical example that got me stuck a while ago:

I was tinkering with a physics based platformer and wanted to implement jump-through platforms. Pretty standard mechanic. However, this requires me to disable the physics of the platform for the player, while possible enemies should still collide with the platform so they don't fall through. I have not found a way to solve this.

**Comment by Smaadz** [comment]
Would love to have this added. I generally avoid using the physics plugin solely because of this.

**Comment by Coolguyzv** [comment]
> Some possible 3D implementations of using physics, if objects are on different Z elevations, they are technically overlapping and therefore collide, even if you do the complex stuff of detecting whether they are on similar Z elevations and such.

This is why I want this feature from the first place as soon as possible. My current solution, as the OP said, is duplicating the object multiple times, and often 10+ different objects have 10 different clones. I often have to make unnecessary families to organize objects by clone so I can filter them in the event sheet, which is a nightmare. I don't know why it's hard to implement by instance and treat each instance as a completely different object to work, I guess.

But I wish physics collisions worked like Solid filters do. It would be easy to make objects interact with different collisions by tags with no effort.

**Comment by Wackytoaster** [comment]
Still hoping to see this happen in 2025. :I

**Comment by Fib100** [comment]
It's definitely possible in Box2d. These parameters just need to be exposed in C3 editor.
https://box2d.org/documentation/group__shape.html#structb2_filter
